buildSrc/src/main/java/org/elasticsearch/gradle/clusterformation/ClusterformationPlugin.java

```diff
@@ -0,0 +1,90 @@
+ */
+package org.elasticsearch.gradle.clusterformation;
+
+import groovy.lang.Closure;
+import org.elasticsearch.GradleServicesAdapter;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.plugins.ExtraPropertiesExtension;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ClusterformationPlugin implements Plugin<Project> {
+
+    public static final String LIST_TASK_NAME = "listElasticSearchClusters";
+    public static final String EXTENSION_NAME = "elasticSearchClusters";
+
+    private final Logger logger =  Logging.getLogger(ClusterformationPlugin.class);
+
+    @Override
+    public void apply(Project project) {
+        NamedDomainObjectContainer<? extends ElasticsearchConfiguration> container = project.container(
+            ElasticsearchNode.class,
+            (name) -> new ElasticsearchNode(name, GradleServicesAdapter.getInstance(project))
+        );
+        project.getExtensions().add(EXTENSION_NAME, container);
+
+        Task listTask = project.getTasks().create(LIST_TASK_NAME);
+        listTask.setGroup("ES cluster formation");
+        listTask.setDescription("Lists all ES clusters configured for this project");
+        listTask.doLast((Task task) ->
+            container.forEach((ElasticsearchConfiguration cluster) ->
+                logger.lifecycle("   * {}: {}", cluster.getName(), cluster.getDistribution())
+            )
+        );
+
+        Map<Task, List<ElasticsearchConfiguration>> taskToCluster = new HashMap<>();
+
+        // register an extension for all current and future tasks, so that any task can declare that it wants to use a
+        // specific cluster.
+        project.getTasks().all((Task task) ->
+            task.getExtensions().findByType(ExtraPropertiesExtension.class)
+            .set(
+                "useCluster",
+                new Closure<Void>(this, this) {
+                    public void doCall(ElasticsearchConfiguration conf) {
+                        taskToCluster.computeIfAbsent(task, k -> new ArrayList<>()).add(conf);
+                    }
+                })
+        );
+
+        // we need to claim all the clusters before starting executing
+        project.getGradle().getTaskGraph().whenReady(taskExecutionGraph ->
+            taskExecutionGraph.getAllTasks()
+                .forEach(task ->
+                    taskToCluster.getOrDefault(task, new ArrayList<>()).forEach(ElasticsearchConfiguration::claim)
+                )
+        );
+
+        // create the listener to start the clusters on-demand and terminate when no longer claimed.
+        project.getGradle().addListener(
+            new ClusterFormationTaskExecutionListener(Collections.unmodifiableMap(taskToCluster))
```

Comment:
This suffers from the same concern I had in my previous review: by using a separate class, it increases the complexity of what must be understood by someone reading this code.

Reply:
I don't agree that the listener class is more complex than hooking on the task graph and doFirst, I would even argue that the listener interface is simpler because it follows a well understood design pattern. To me  this fits better than a plugin implemented in java. 

It does have all the advantages, if a task is skipped, no actions will be run for that task, so  `beforeActions` is never called and thus the cluster is not started.

The real tie breaker for me is that it also allows for tasks to have their own doFirst that interacts with the cluster, e.x. a testing task that does some setup before running the test, like indexing ( I think docs tests do this ). Keep in mind that there won't be a separate start cluster task. If we start the cluster in `doFrist` there will be no way for build scripts to add any actions that run after the cluster is started but before the regular task actions. 

W.r.t immutability, the listener gets an immutable map and the configurations themselves lock-down once the cluster is running. This will guarantee that we don't change configuration in the execution phase.

Reply:
> The real tie breaker for me is that it also allows for tasks to have their own doFirst

My proposed solution still allows this. The doFirst is not added until the task graph is resolved, after configuration is complete. So the doFirst added is actually the very first doFirst, before any configured doFirst blocks.

> the configurations themselves lock-down once the cluster is running

But they don't actually, because the "claimed" tracking is done there, which is always mutable.

> I would even argue that the listener interface is simpler because it follows a well understood design pattern

All of these are hooks that allow executing a block of code before/after a task is run. They have different nuances, but they are all the same design pattern. Additionally, you could still use the listener, but I am suggesting it can be created as either an inner class, or an anonymous class, inside the plugin, so that the logic and state are all together.